File: src/sigma/bridge.py
"""Native macOS file dialogs, exposed to the interface through pywebview.

The web layer cannot open a real Finder dialog, and `<input type="file">` only
yields an upload — never a path the app can keep writing to. It is also
unreliable inside WKWebView. So the dialogs are opened by pywebview and the
chosen path is handed back to JavaScript as ``window.pywebview.api``.

Every method returns a plain dict so the frontend can treat a cancelled dialog
(``{"path": None}``) the same way as a successful one.
"""

from __future__ import annotations

from pathlib import Path
from typing import Any

DIALOG_FILTER = ("Base de datos Sigma (*.db)", "*.db")


class Bridge:
    """Methods callable from JavaScript via ``window.pywebview.api``."""

    def __init__(self) -> None:
        self._window: Any = None

    def attach(self, window: Any) -> None:
        self._window = window

    # -- Dialogs ------------------------------------------------------------

    def choose_database(self) -> dict[str, str | None]:
        """Ask for an existing ``.db`` file to open."""
        import webview

        result = self._dialog(webview.OPEN_DIALOG, allow_multiple=False)
        return {"path": result}

    def choose_new_database(self) -> dict[str, str | None]:
        """Ask where to save a new database.

        Defaults to the user's Google Drive folder when one is present, since
        keeping the file in a synced folder is the whole point of choosing it.
        """
        import webview

        result = self._dialog(
            webview.SAVE_DIALOG,
            directory=str(_default_folder()),
            save_filename="finanzas.db",
        )
        return {"path": result}

    def open_releases(self) -> dict[str, bool]:
        """Open the downloads page in the default browser.

        Takes no URL on purpose: JavaScript inside the window must not be able
        to hand `open` an arbitrary target. Following the link inside the
        window would replace the app with a web page and leave no way back.
        """
        import subprocess

        from sigma.updates import RELEASES_PAGE

        subprocess.run(["open", RELEASES_PAGE], check=False)
        return {"ok": True}

    def quit(self) -> dict[str, bool]:
        """Close the window, which ends the process.

        Used after staging an update: the script that swaps the bundle is
        already waiting for this process to go away. Closing happens on its own
        thread so the call from JavaScript can return first.
        """
        import threading

        if self._window is not None:
            threading.Timer(0.1, self._window.destroy).start()
        return {"ok": True}

    def reveal(self, path: str) -> dict[str, bool]:
        """Show a file in Finder."""
        import subprocess

        target = Path(path).expanduser()
        if not target.exists():
            return {"ok": False}
        subprocess.run(["open", "-R", str(target)], check=False)
        return {"ok": True}

    # -- Internals ----------------------------------------------------------

    def _dialog(self, dialog_type: int, **kwargs: Any) -> str | None:
        if self._window is None:
            return None
        result = self._window.create_file_dialog(
            dialog_type, file_types=(DIALOG_FILTER[0], "Todos los archivos (*.*)"), **kwargs
        )
        if not result:
            return None
        # OPEN_DIALOG returns a sequence, SAVE_DIALOG a single path.
        return str(result[0]) if isinstance(result, (list, tuple)) else str(result)


def _default_folder() -> Path:
    """Google Drive's local folder if it exists, otherwise the home directory."""
    candidates = [
        Path.home() / "Library" / "CloudStorage",
        Path.home() / "Google Drive",
        Path.home() / "Documents",
    ]
    for candidate in candidates:
        if candidate.is_dir():
            return candidate
    return Path.home()

File: web/src/lib/bridge.ts
/**
 * Access to the native macOS file dialogs exposed by `sigma/bridge.py`.
 *
 * pywebview injects `window.pywebview.api` shortly after the page loads, so
 * calls wait for it rather than assuming it is already there. In the browser
 * during `npm run dev` there is no bridge at all; `isAvailable` lets the
 * interface offer a plain text field instead of a dead button.
 */

interface PywebviewApi {
  choose_database(): Promise<{ path: string | null }>;
  choose_new_database(): Promise<{ path: string | null }>;
  reveal(path: string): Promise<{ ok: boolean }>;
  open_releases(): Promise<{ ok: boolean }>;
  quit(): Promise<{ ok: boolean }>;
}

declare global {
  interface Window {
    pywebview?: { api?: PywebviewApi };
  }
}

const READY_TIMEOUT_MS = 3000;

function api(): PywebviewApi | null {
  return window.pywebview?.api ?? null;
}

export function isAvailable(): boolean {
  return api() !== null;
}

async function waitForBridge(): Promise<PywebviewApi | null> {
  const deadline = Date.now() + READY_TIMEOUT_MS;
  while (Date.now() < deadline) {
    const bridge = api();
    if (bridge) return bridge;
    await new Promise((resolve) => setTimeout(resolve, 50));
  }
  return null;
}

/** Ask for an existing database file. Resolves to `null` if cancelled. */
export async function chooseDatabase(): Promise<string | null> {
  const bridge = await waitForBridge();
  if (!bridge) return null;
  return (await bridge.choose_database()).path;
}

/** Ask where to put a new database file. Resolves to `null` if cancelled. */
export async function chooseNewDatabase(): Promise<string | null> {
  const bridge = await waitForBridge();
  if (!bridge) return null;
  return (await bridge.choose_new_database()).path;
}

export async function revealInFinder(path: string): Promise<void> {
  const bridge = api();
  if (bridge) await bridge.reveal(path);
}

/**
 * Close the app. Used after an update is staged: the script that replaces the
 * bundle is waiting for this process to exit, and reopens Sigma afterwards.
 */
export async function quit(): Promise<void> {
  const bridge = api();
  if (bridge) await bridge.quit();
}

/**
 * Show the downloads page in the browser. Inside the app window it has to be
 * the native `open`: navigating there would replace the interface with a web
 * page and there is no back button. `window.open` is the fallback for
 * `npm run dev`, where there is no bridge.
 */
export async function openReleases(url: string): Promise<void> {
  const bridge = api();
  if (bridge) {
    await bridge.open_releases();
    return;
  }
  window.open(url, '_blank', 'noopener');
}
